typed-racket-lib/typed-racket/base-env/prims-struct.rkt

```diff
@@ -176,6 +177,25 @@
                              (second (build-struct-names #'nm.name null #t #t))))
                 . opts))]))
 
+(define-for-syntax (maybe-get-prop-proc-ty prop-names prop-vals)
+  (for/first ([nm (in-list prop-names)]
+              [val (in-list prop-vals)]
+              #:when (free-identifier=? nm #'prop:procedure))
+    (syntax-parse val
+      #:literals (struct-field-index)
+      #:datum-literals (lambda : λ ann lambda:)
+      [((~or lambda λ) ((_ : ty) ...) : ret-ty _)
+       (syntax/loc val
+         (-> ty ... ret-ty))]
+      [(ann ((~or lambda λ lambda:) formals . rst) ty)
+       #'ty]
+      [_:exact-nonnegative-integer
+       val]
+      [(struct-field-index fld)
+       #'fld]
+      [_
+       (tc-error/stx val "expected: a nonnegative integer literal or an annotated lambda")])))
```

Comment:
Delay this error until `tc-struct` (consider programs in `typed/racket/no-check`).